### PR TITLE
Improve export and encoding

### PR DESCRIPTION
**Pull request by turbolent**
## Description

- Add support for encoding attachments to JSON-CDC codec
- Handle `nil` values when encoding JSON-CDC. This occurs for example in the result of a cyclic reference
- Fix encoding of composite values which have attachments attached to them in JSON-CDC encoding 
- In general, test that script return values in test cases can also be encoded by JSON-CDC. 
  It would be nice to do that for CCF as well, once it is ready

cc @fxamacker 

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: encoding/json/encode.go [hunk 1]
```diff
@@ -618,10 +627,14 @@ func prepareComposite(kind, id string, fieldTypes []cadence.Field, fields []cade
 	compositeFields := make([]jsonCompositeField, len(fields))
 
 	for i, value := range fields {
-		fieldType := fieldTypes[i]
+		var name string
+		if i < len(fieldTypes) {
+			fieldType := fieldTypes[i]
+			name = fieldType.Identifier
+		}
```
Comment (SupunS) [inline encoding/json/encode.go:634]: We already check this at the top of the function, so wouldn't need to do it here again?

Comment (turbolent) [reply encoding/json/encode.go:634]: I'll add comments to clarify

File: encoding/json/encode.go [hunk 2]
```diff
@@ -605,8 +610,12 @@ func prepareEnum(v cadence.Enum) jsonValue {
 	return prepareComposite(enumTypeStr, v.EnumType.ID(), v.EnumType.Fields, v.Fields)
 }
 
+func prepareAttachment(v cadence.Attachment) jsonValue {
+	return prepareComposite(attachmentTypeStr, v.AttachmentType.ID(), v.AttachmentType.Fields, v.Fields)
+}
+
 func prepareComposite(kind, id string, fieldTypes []cadence.Field, fields []cadence.Value) jsonValue {
-	if len(fieldTypes) != len(fields) {
+	if len(fields) < len(fieldTypes) {
```
Comment (turbolent) [inline encoding/json/encode.go:618]:
```suggestion
	// Ensure there are _at least _ as many field values as field types.
	// There might be more field values in the case of attachments.
	if len(fields) < len(fieldTypes) {
```

File: encoding/json/encode.go [hunk 3]
```diff
@@ -618,10 +627,14 @@ func prepareComposite(kind, id string, fieldTypes []cadence.Field, fields []cade
 	compositeFields := make([]jsonCompositeField, len(fields))
 
 	for i, value := range fields {
-		fieldType := fieldTypes[i]
+		var name string
+		if i < len(fieldTypes) {
```
Comment (turbolent) [inline encoding/json/encode.go:631]:
```suggestion
		// Provide the field name, if the field type is available.
		// In the case of attachments, they are provided as field values,
		// but there is no corresponding field type.
		if i < len(fieldTypes) {
```

File: encoding/json/encode.go (hunk 1 above)
Comment (turbolent) [reply encoding/json/encode.go:639]: Explained it in [a9b2f77](https://github.com/onflow/cadence/pull/2517/commits/a9b2f77ed8350778c742eb9a69cdf6118f028e83)

**Pull request merged**